File: examples/lua/src/server/todos.server.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local RunService = game:GetService("RunService")

local remotes = require(ReplicatedStorage.Shared.remotes)

local todos: { string } = { "Milk", "Eggs" }
local changed = false

remotes.addTodo:connect(function(player, todo)
	table.insert(todos, todo)
	changed = true
end)

remotes.removeTodo:connect(function(player, todo)
	table.remove(todos, table.find(todos, todo) or -1)
	changed = true
end)

remotes.getTodos:onRequest(function(player)
	return todos
end)

RunService.Heartbeat:Connect(function()
	if changed then
		changed = false
		remotes.todosChanged:fireAll(todos)
		print("New todos:", table.concat(todos, ", "))
	end
end)

File: examples/ts/src/server/todos.server.ts
import { RunService } from "@rbxts/services";
import { remotes } from "shared/remotes";

const todos: string[] = ["Milk", "Eggs"];
let changed = false;

remotes.addTodo.connect((player, todo) => {
	todos.push(todo);
	changed = true;
});

remotes.removeTodo.connect((player, todo) => {
	todos.remove(todos.indexOf(todo));
	changed = true;
});

remotes.getTodos.onRequest((player) => {
	return todos;
});

RunService.Heartbeat.Connect(() => {
	if (changed) {
		changed = false;
		remotes.todosChanged.fireAll(todos);
		print(`New todos: ${todos.join(", ")}`);
	}
});
